File: C#/CustomSortString.cs
// Problem: https://leetcode.com/problems/custom-sort-string/submissions/
// Author: Dung Nguyen Tien (Chris)

public class Solution {
    public string CustomSortString(string S, string T) {
        var dict = new Dictionary<char, int>();
        
        var i = 0;
        foreach (var c in S) {
            if (!dict.ContainsKey(c)) {
                dict.Add(c, i++);
            }
        }
        
        var customSortArray = T.ToCharArray();
        Array.Sort(customSortArray, (x, y) => {
            var ix = dict.ContainsKey(x) ? dict[x] : int.MaxValue;
            var iy = dict.ContainsKey(y) ? dict[y] : int.MaxValue;
            
            return ix - iy;
        });
        
        T = new string(customSortArray);
        return T;
    }
}

File: Javascript/CustomSortString.js
/**
 * Problem: https://leetcode.com/problems/custom-sort-string/submissions/
 * Author: Dung Nguyen Tien (Chris)
 * @param {string} S
 * @param {string} T
 * @return {string}
 */
var customSortString = function(S, T) {
    return T.split('')
            .sort(function(a, b) { 
                var ia = S.indexOf(a) >= 0 ? S.indexOf(a) : Number.MAX_SAFE_INTEGER;
                var ib = S.indexOf(b) >= 0 ? S.indexOf(b) : Number.MAX_SAFE_INTEGER
                return ia - ib;
                //return (S.indexOf(a) | 0) - (S.indexOf(b) | 0);
            })
            .join('');
};